doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,159 @@
+
+> **Note:** Some details of a `get` operation are defined in the [contract with backend plugins](./backend-plugins.md).
+
+Properties of `kdbGet()`:
+
+- After calling `kdbGet (kdb, ks, parentKey)`, the KeySet `ks` will contain _all keys_ (including their values) that are stored in _any backend_ with a mountpoint that is _below `parentKey`_.
+- After calling `kdbGet (kdb, ks, parentKey)`, below `parentKey` the KeySet `ks` will _only_ contain keys that are stored in a backend.
+- The KeySet `ks` may contain other keys not below `parentKey`.
+  These keys fall into one of two categories:
+  1. Keys that are stored in a backend that is not below `parentKey`.
+     `kdbGet()` may decide that it is more efficient (e.g. because of a cache) to return more keys than requested.
+  2. Keys that were already present in `ks` when `kdbGet()` was called.
+
+In simpler terms, this means to the caller it looks as if `kdbGet()` had removed all keys below `parentKey` from `ks` and then loaded the data from the backends.
+Below `parentKey` the KeySet `ks` correctly represents the state of the KDB.
+For the rest of `ks` there are no such guarantees.
+
+TODO: cache correct? notifications? (just call at the end?)
+
+The basic flow of this operation is:
+
+1. Determine the backends needed to read all keys below `parentKey`
+2. Run the `init` phase on all the backends that haven't been initialized
+3. Run the `resolver` phase on all backends
+4. From now on ignore all backends, which indicated that there is no update.
+5. If all backends are now ignored, **return**.
+6. If a global cache plugin is enabled:
+   Ask the global cache plugin for the cache times for all backends.
+7. If all backends have an existing cache entry:
+   Run the `cachecheck` phase on all backends
+8. If all backends indicated the cache is still valid:
+   Ask the global cache plugin for the cached data and **return**.
+9. Run the `prestorage` and `storage` phase on all backends.
+10. Run the `poststorage` phase of all `spec:/` backends.
+11. Merge the data from all backends
+12. If enabled, run the `gopts` plugin.
+13. Run the `spec` plugin (to copy metakeys).
+14. Split data back into individual backends.
+15. Run the `poststorage` phase for all backends.
+16. Merge the data from all backends.
+17. If a global cache plugin is enabled, update cache.
+    Then **return**.
+
+> **Note:** In case of error, we abort immediately, restore `ks` to its original state and return.
+
+Influence of namespaces:
+
+- cascading and `meta:/` keys are always illegal in `ks` (should be enforced via different KeySet types)
+- `spec:/` backends only go through `init`, `resolver`, `cache`, `presetstorage` and `storage` phases as normal, but their `poststorage` phase is called earlier.
+- `dir:/`, `user:/` and `system:/` go through all phases as described above.
+- `proc:/` mountpoints go through all the phases as described above, but they are not stored in the cache.
```

Comment:
The interaction of `proc:/` mountpoints and the cache is not quite clear yet. The easiest solution would be to just ignore the cache (no load, no store), if a `proc:/` mountpoint is used. The problem is that `gopts` kind of acts like a `proc:/` mountpoint and an incompatibility between cache and `gopts` would be bad.

I think it should be safe to just load the cache run `prestorage` and `storage` for `proc:/` then call `gopts` and `spec` and finally run `poststorage` for `proc:/`. The assumption here is that a change in `proc:/` keys can never affect other keys. For `proc:/` mountpoints and `gopts` this is the case (both should only write to `proc:/`, needs to be enforced). The question is what we do with `spec`. The current implementation shouldn't have a problem AFAIK, but IIRC we wanted to generate more metakeys (e.g. `array`) there might be some edge cases there. I also haven't thought about notifications at all yet.

Reply:
This was mostly I note to myself (I had already forgotten, about the cache + `proc:/` problem). I'm pretty sure the strategy of "if cache-hit -> then load `proc:/` in isolation" should work fine. If you have no issues with this either then there is nothing to discuss.